core/kernel/dt_driver_test.c

```diff
@@ -0,0 +1,529 @@
+
+	if (res == TEE_ERROR_DEFER_DRIVER_INIT &&
+	    dt_test_state.probe_deferral == DEFAULT) {
+		/* We expect at least a probe deferral */
+		dt_test_state.probe_deferral = IN_PROGRESS;
+	}
+
+	return res;
+}
+
+static const struct dt_device_match dt_test_consumer_match_table[] = {
+	{ .compatible = "linaro,dt-test-consumer", },
+	{ }
+};
+
+DEFINE_DT_DRIVER(dt_test_consumer_driver) = {
+	.name = "dt-test-consumer",
+	.match_table = dt_test_consumer_match_table,
+	.probe = dt_test_consumer_probe,
+};
+
+static TEE_Result dt_test_crypt_consumer_probe(const void *fdt __unused,
+					       int node __unused,
+					       const void *compat_data __unused)
+{
+	TEE_Result res = dt_driver_get_crypto();
+	uint8_t __maybe_unused byte = 0;
+
+	if (res == TEE_ERROR_DEFER_DRIVER_INIT &&
+	    dt_test_state.crypto_dependencies == DEFAULT) {
+		/* We expect to be deferred */
+		dt_test_state.crypto_dependencies = IN_PROGRESS;
+	}
+
+	if (res)
+		return res;
+
+	if (dt_test_state.crypto_dependencies == DEFAULT) {
+		EMSG("Test expects at least a driver probe deferral");
+		dt_test_state.crypto_dependencies = FAILED;
+		return TEE_ERROR_GENERIC;
+	}
+
+	if (crypto_rng_read(&byte, sizeof(byte))) {
+		dt_test_state.crypto_dependencies = FAILED;
+		return TEE_ERROR_GENERIC;
+	}
+
+	dt_test_state.crypto_dependencies = SUCCESS;
+	return TEE_SUCCESS;
+}
+
+static const struct dt_device_match dt_test_crypt_consumer_match_table[] = {
+	{ .compatible = "linaro,dt-test-crypt-consumer", },
+	{ }
+};
+
+DEFINE_DT_DRIVER(dt_test_consumer_driver) = {
+	.name = "dt-test-crypt-consumer",
+	.match_table = dt_test_crypt_consumer_match_table,
+	.probe = dt_test_crypt_consumer_probe,
+};
+
+#ifdef CFG_DRIVERS_CLK
+#define DT_TEST_CLK_COUNT		2
+
+#define DT_TEST_CLK0_BINDING_ID		3
+#define DT_TEST_CLK1_BINDING_ID		7
+
+static const char *dt_test_clk_name[DT_TEST_CLK_COUNT] = {
+	"dt_test-clk3",
+	"dt_test-clk7",
+};
+
+/* Emulating a clock does not require operators */
+static const struct clk_ops dt_test_clock_provider_ops;
+
+static struct clk *dt_test_get_clk(struct dt_driver_phandle_args *args,
+				   void *data, TEE_Result *res)
+{
+	struct clk *clk_ref = data;
+	struct clk *clk = NULL;
+
+	if (args->args_count != 1) {
+		*res = TEE_ERROR_BAD_PARAMETERS;
+		return NULL;
+	}
+
+	switch (args->args[0]) {
+	case DT_TEST_CLK0_BINDING_ID:
+		clk = clk_ref;
+		break;
+	case DT_TEST_CLK1_BINDING_ID:
+		clk = clk_ref + 1;
+		break;
+	default:
+		EMSG("Unexpected binding ID %"PRIu32, args->args[0]);
+		*res = TEE_ERROR_BAD_PARAMETERS;
+		return NULL;
+	}
+
+	DT_TEST_MSG("Providing clock %s", clk_get_name(clk));
+
+	*res = TEE_SUCCESS;
+	return clk;
+}
+
+static TEE_Result dt_test_clock_provider_probe(const void *fdt, int node,
+					       const void *compat_data __unused)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct clk *clk = NULL;
+
+	DT_TEST_MSG("Register clocks");
+
+	clk = dt_test_alloc(DT_TEST_CLK_COUNT * sizeof(*clk));
+	if (!clk)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
```

Comment:
``` C
for (n = 0; n < DT_TEST_CLK_COUNT; n++) {
...
```
should avoid these hard coded indexes below.

Reply:
ok